src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -475,6 +482,39 @@ entails SolverOptions{..} constraint context hints =
       pure (arg1, arg2, srcTypeLevelString (mkString $ h' <> t'))
     consSymbol _ _ _ = Nothing
 
+    solveIsNat :: [SourceType] -> Maybe [TypeClassDict]
+    solveIsNat [TypeLevelNat ann nat] = Just [TypeClassDictionaryInScope Nothing 0 (IsNatInstance nat) [] C.IsNat [] [] [TypeLevelNat ann nat] Nothing Nothing]
+    solveIsNat _ = Nothing
+
+    solveNatAdd :: [SourceType] -> Maybe [TypeClassDict]
+    solveNatAdd [arg0, arg1, arg2] = do
+      (arg0', arg1', arg2') <- addNats arg0 arg1 arg2
+      let args' = [arg0', arg1', arg2']
+      pure [TypeClassDictionaryInScope Nothing 0 EmptyClassInstance [] C.NatAdd [] [] args' Nothing Nothing]
+    solveNatAdd _ = Nothing
+
+    addNats :: SourceType -> SourceType -> SourceType -> Maybe (SourceType, SourceType, SourceType)
+    addNats arg0@(TypeLevelNat _ l) arg1@(TypeLevelNat _ r) _ = pure (arg0, arg1, srcTypeLevelNat (l + r))
+    addNats arg0@(TypeLevelNat _ l) _ arg2@(TypeLevelNat _ r) = pure (arg0, srcTypeLevelNat (l + r), arg2)
+    addNats _ arg1@(TypeLevelNat _ l) arg2@(TypeLevelNat _ r) = pure (srcTypeLevelNat (l + r), arg1, arg2)
```

Comment:
I may be misunderstanding this part of the code checker. However, it seems that the first case of `addNats` is correctly states, "Given `l r o`, `o == l + r`".

However, the second case seems to say, "Given `l r o`, `r == l + o` rather than `r == o - l`, which should only hold if the resulting `r` is a natural number. Likewise for the third case.

If you dropped the second and third case, then you'd need to drop the second and third functional dependency of the `AddNat` type class.

Reply:
Ah, this was left-over from when I was testing stuff. I'll comment this out in the meantime until there's a determinate implementation for `Sub`.